homeassistant/components/shelly/config_flow.py

```diff
@@ -48,33 +52,69 @@ async def async_step_user(self, user_input=None):
         errors = {}
         if user_input is not None:
             try:
-                info = await self._async_get_info(user_input["host"])
+                self.host = user_input[CONF_HOST]
+                info = await self._async_get_info(self.host)
             except HTTP_CONNECT_ERRORS:
                 errors["base"] = "cannot_connect"
             except Exception:  # pylint: disable=broad-except
                 _LOGGER.exception("Unexpected exception")
                 errors["base"] = "unknown"
 
             else:
+                await self.async_set_unique_id(info["mac"])
+                self._abort_if_unique_id_configured()
                 if info["auth"]:
-                    return self.async_abort(reason="auth_not_supported")
+                    return await self.async_step_credentials()
 
                 try:
-                    device_info = await validate_input(self.hass, user_input)
+                    device_info = await validate_input(self.hass, self.host, user_input)
                 except HTTP_CONNECT_ERRORS:
                     errors["base"] = "cannot_connect"
                 except Exception:  # pylint: disable=broad-except
                     _LOGGER.exception("Unexpected exception")
                     errors["base"] = "unknown"
                 else:
-                    await self.async_set_unique_id(device_info["mac"])
                     return self.async_create_entry(
-                        title=device_info["title"] or user_input["host"],
+                        title=device_info["title"] or self.host,
                         data=user_input,
                     )
 
         return self.async_show_form(
-            step_id="user", data_schema=DATA_SCHEMA, errors=errors
+            step_id="user", data_schema=HOST_SCHEMA, errors=errors
+        )
+
+    async def async_step_credentials(self, user_input=None):
+        """Handle the credentials step."""
+        errors = {}
+        schema = None
+        if user_input is not None:
+            try:
+                device_info = await validate_input(self.hass, self.host, user_input)
+            except aiohttp.ClientResponseError as error:
+                if error.status == 401:
+                    errors["base"] = "invalid_auth"
+                else:
+                    errors["base"] = "cannot_connect"
+            except Exception:  # pylint: disable=broad-except
+                _LOGGER.exception("Unexpected exception")
+                errors["base"] = "unknown"
+            else:
+                return self.async_create_entry(
+                    title=device_info["title"] or self.host,
+                    data={**user_input, **{CONF_HOST: self.host}},
```

Comment:
We don't need to create a dict an unpack it inside the new dict. Just list the host items after a comma in the new dict.

Reply:
```suggestion
                    data={**user_input, CONF_HOST: self.host},
```